### PR TITLE
Add Rectangle and test

### PR DESCRIPTION
**Pull request by mar-celine**

### CODE REVIEW

File: src/Domain/DomainCreators/Rectangle.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    const OptionContext& /*context*/)
+    : lower_xy_(lower_xy),
+      upper_xy_(upper_xy),
+      is_periodic_in_xy_(is_periodic_in_xy),
+      initial_refinement_level_xy_(initial_refinement_level_xy),
+      initial_number_of_grid_points_in_xy_(
+          initial_number_of_grid_points_in_xy) {}
+
+Domain<2, Frame::Inertial> Rectangle::create_domain() const {
+  using affine_map = CoordinateMaps::AffineMap;
```
Comment (osheamonn) [inline src/Domain/DomainCreators/Rectangle.cpp]: Since this is being used as a type as opposed to a metavariable I'd recommend `using AffineMap`

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include "Options/Options.hpp"
+
+namespace DomainCreators {
+
+/// Create a 2D Domain consisting of a single Block.
+class Rectangle : public DomainCreator<2, Frame::Inertial> {
+ public:
+  struct LowerBound {
+    using type = std::array<double, 2>;
+    static constexpr OptionString_t label = {"LowerBound"};
```
Comment (wthrowe) [inline src/Domain/DomainCreators/Rectangle.hpp]: Remove `label` from all the option structs.

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    static constexpr OptionString_t help = {
+        "Sequence of [x,y] for lower bounds."};
+    static size_t size() { return 2; }
```
Comment (wthrowe) [inline src/Domain/DomainCreators/Rectangle.hpp]: Remove `size` from all the option structs.

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+          {{Direction<2>::lower_xi()},
+           {Direction<2>::upper_xi()},
+           {Direction<2>::lower_eta()},
+           {Direction<2>::upper_eta()}}},
+      make_vector(make_coordinate_map<Frame::Logical, Frame::Inertial>(
+          affine_map_2d{affine_map{-1., 1., lower_bound[0], upper_bound[0]},
+                        affine_map{-1., 1., lower_bound[1], upper_bound[1]}})));
+}
+
+void test_periodic_rectangle(
```
Comment (mar-celine) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp]: need to test other periodic boundaries as well (only periodic in x, periodic in x and y, etc)

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+/// Defines class Rectangle.
+
+#pragma once
+
+#include "Domain/DomainCreators/DomainCreator.hpp"
+#include "Options/Options.hpp"
+
+namespace DomainCreators {
+
+/// Create a 2D Domain consisting of a single Block.
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.hpp]: Doxygen group

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Domain.DomainCreators.Rectangle", "[Domain][Unit]") {
+  const std::array<size_t, 2> grid_points{{4, 6}}, num_elements{{3, 2}};
+  const std::array<double, 2> lower_bound{{-1.2, 3.0}}, upper_bound{{0.8, 5.0}};
+
+  const DomainCreators::Rectangle rectangle{lower_bound, upper_bound,
+                                            std::array<bool, 2>{{false, false}},
+                                            num_elements, grid_points};
+
```
Comment (kidder) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp]: remove blank line

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+  test_rectangle(rectangle, lower_bound, upper_bound, grid_points,
+                 num_elements);
+  const DomainCreators::Rectangle periodic_x_rectangle{
```
Comment (kidder) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp]: add blank line

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+      lower_bound, upper_bound, std::array<bool, 2>{{true, false}},
+      num_elements, grid_points};
+  test_periodic_x_rectangle(periodic_x_rectangle, lower_bound, upper_bound,
+                            grid_points, num_elements);
+  const DomainCreators::Rectangle periodic_y_rectangle{
```
Comment (kidder) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp]: add blank line

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      const OptionContext& context = {});
+
+  Rectangle() = default;
+  Rectangle(const Rectangle&) = delete;
+  Rectangle(Rectangle&&) noexcept = default;
+  Rectangle& operator=(const Rectangle&) = delete;
+  Rectangle& operator=(Rectangle&&) noexcept = default;
+  ~Rectangle() override = default;
+
+  Domain<2, Frame::Inertial> create_domain() const override;
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.hpp]: noexcept the member functions

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    using type = std::array<size_t, 2>;
+    static constexpr OptionString_t help = {
+        "Initial number of grid points in [x,y]."};
+  };
+  using options = tmpl::list<LowerBound, UpperBound, IsPeriodicIn,
+                             InitialRefinement, InitialGridPoints>;
+
+  static constexpr OptionString_t help{"Creates a 2D rectangle."};
+
+  Rectangle(
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.hpp:52]: noexcept

File: src/Domain/DomainCreators/Rectangle.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+namespace DomainCreators {
+
+Rectangle::Rectangle(
+    const typename LowerBound::type lower_xy,
+    const typename UpperBound::type upper_xy,
+    const typename IsPeriodicIn::type is_periodic_in_xy,
+    const typename InitialRefinement::type initial_refinement_level_xy,
+    const typename InitialGridPoints::type initial_number_of_grid_points_in_xy,
+    const OptionContext& /*context*/)
+    : lower_xy_(lower_xy),
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.cpp]: make non-const, use std::move

File: src/Domain/DomainCreators/Rectangle.cpp [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+      make_vector<std::unique_ptr<
+          CoordinateMapBase<Frame::Logical, Frame::Inertial, 2>>>(
+          std::make_unique<
+              CoordinateMap<Frame::Logical, Frame::Inertial, AffineMap2D>>(
+              AffineMap2D{AffineMap{-1., 1., lower_xy_[0], upper_xy_[0]},
+                          AffineMap{-1., 1., lower_xy_[1], upper_xy_[1]}})),
+      std::vector<std::array<size_t, 4>>{{{0, 1, 2, 3}}}, identifications};
+}
+
+std::array<size_t, 2> Rectangle::initial_extents(
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.cpp]: add tests for `ASSERT`s

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+#include "Domain/DomainCreators/Rectangle.hpp"
+#include "Utilities/MakeVector.hpp"
+#include "tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp"
+#include "tests/Unit/Domain/TestDomainHelpers.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+using AffineMap = CoordinateMaps::AffineMap;
+using AffineMap2D = CoordinateMaps::ProductOf2Maps<AffineMap, AffineMap>;
+void test_rectangle(const DomainCreators::Rectangle& rectangle,
```
Comment (nilsdeppe) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp]: simplify all these functions into one, or even just a loop.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Domain/DomainCreators/Rectangle.hpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+/// Defines class Rectangle.
+
+#pragma once
+
+#include "Domain/DomainCreators/DomainCreator.hpp"
+#include "Options/Options.hpp"
+
+namespace DomainCreators {
+
+/// \ingroup DomainCreatorsGroup
```
Comment (nilsdeppe) [inline src/Domain/DomainCreators/Rectangle.hpp:14]: Please properly propagate this change throughout the code base.

File: src/Domain/DomainCreators/Rectangle.hpp (hunk 9 above)
Comment (nilsdeppe) [reply src/Domain/DomainCreators/Rectangle.hpp]: bump

File: src/Domain/DomainCreators/Rectangle.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Domain/DomainCreators/Rectangle.hpp:52]: bump

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Review by kidder — CHANGES_REQUESTED** [review]
fix the DomainCreatorsGroup and squash

**Comment by nilsdeppe** [comment]
Please fix clang-tidy warnings

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+  test_factory_creation<DomainCreator<2, Frame::Inertial>>(
+      "  Rectangle:\n"
+      "    LowerBound: [0,0]\n"
+      "    UpperBound: [1,2]\n"
+      "    IsPeriodicIn: [True,False]\n"
+      "    InitialGridPoints: [3,4]\n"
+      "    InitialRefinement: [2,3]\n");
+}
+
+// [[OutputRegex, index = 1]]
```
Comment (nilsdeppe) [inline tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp:110]: Check more of the ASSERT message, ie the condition that failed. I believe that is printed.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp (hunk 15 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/DomainCreators/Test_Rectangle.cpp:110]: I changed my mind on this: these functions are going away anyway, so all is good.

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
